### PR TITLE
Remove inheritance on pathlib

### PR DESCRIPTION
**Pull request by qubixes**
And reimplement some of its methods.

### CODE REVIEW

**Comment by qubixes** [comment]
@chStaiger I have added some tests, but I have also removed some tests, since I'm not sure exactly what you are trying to do, especially with the windows_paths and their joins. I think for now it is more sensible if the IrodsPath assumes good/valid input for all its functions, meaning the inputs should be posix paths as well. Any wrangling you want to do with windows paths should be done on the data operations side I think.

File: ibridges/irodsconnector/session.py [hunk 1]
```diff
@@ -37,10 +37,30 @@ def __init__(self, irods_env: Optional[dict] = None, irods_env_path: Optional[st
...
         self._password = password
         self._irods_env = irods_env
         self._irods_env_path = irods_env_path
         self._irods_session = self.connect()
+        if irods_home is not None:
+            self._irods_env["irods_home"] = irods_home
+        if "irods_home" not in self._irods_env:
+            self._irods_env["irods_home"] = '/'+self.zone+'/home/'+self.username
+
+    def home(self) -> str:
```
Comment (chStaiger) [inline ibridges/irodsconnector/session.py:50]: Can we also make a home setter? When you are connected to iRODS and want to change your current working directory (irods_home) and you do not want to create a new session.

File: ibridges/irodsconnector/session.py [hunk 2]
```diff
@@ -17,7 +17,7 @@ class Session:
     """
 
     def __init__(self, irods_env: Optional[dict] = None, irods_env_path: Optional[str] = None,
-                 password: Optional[str] = None):
+                 password: Optional[str] = None, irods_home: Optional[str] = None):
         """ iRODS authentication with Python client.
 
         Parameters
```
Comment (chStaiger) [inline ibridges/irodsconnector/session.py:23]: Add hte parameter irods_home. Set to overwrite the "irods_home" value in your irods_environment.json.

File: ibridges/utils/path.py [hunk 3]
```diff
@@ -45,25 +76,27 @@ def remove(self):
...
         """
         try:
             return session.irods_session.collections.create(str(coll_path))
-        except irods.exception.CUT_ACTION_PROCESSED_ERR:
-            raise(irods.exception.CUT_ACTION_PROCESSED_ERR('iRODS server forbids action.'))
+        except irods.exception.CUT_ACTION_PROCESSED_ERR as exc:
+            raise irods.exception.CUT_ACTION_PROCESSED_ERR(
+                "While creating collection '{coll_path}': iRODS server forbids action.") from exc
 
     def rename(self):
```
Comment (chStaiger) [inline ibridges/utils/path.py:101]: The method is not implemented yet. Maybe we can print that to the prompt or throw a warning.

File: ibridges/utils/path.py [hunk 4]
```diff
@@ -82,23 +115,11 @@ def dataobject_exists(self) -> bool:
...
-
-    def home(self):
-        """
-        If the session environment defines an 'irods_home', checks if this path exists 
-        and returns the path.
-        If 'irods_home' is not defined, returns the path /zone/home
-        """
+        return self.dataobject_exists() or self.collection_exists()
 
     def walk(self, depth: int):
```
Comment (chStaiger) [inline ibridges/utils/path.py:124]: The function is not implemented and does nothing. Can we throw a warning?

File: tests/test_irodspath.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    zone = "testzone"
+    user = "testuser"
+    host = "test.host.nl"
+    server_version = "test_version"
+    port = 9876
+
+    def home(self):
+        return "/"+self.zone+"/home/"+self.user
+
+    def cwd(self):
```
Comment (chStaiger) [inline tests/test_irodspath.py:18]: We do not need the cwd any longer, right?

**Review by chStaiger — CHANGES_REQUESTED** [review]
Thanks a lot. This class makes it much easier to to create upload and download paths. Also having tests is really nice.
I added a small tutorial how to work with the new class and I think we can clean the tests up and delete the joinpath section with Windows and Posix paths.

However, there is a bit of a clean-up to do. The data_ops.upload does not work anylonger:

```
data_ops.upload(local_path, irods_path, overwrite = True)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [8], in <cell line: 8>()
      5 f.close()
      7 # upload the file to our collection
----> 8 data_ops.upload(local_path, irods_path, overwrite = True)
     10 # list again the data object in te collection
     11 print('data objects\t', coll.data_objects)

File ~/git-repos/iBridges/ibridges/irodsconnector/data_operations.py:286, in DataOperations.upload(self, local_path, irods_path, overwrite, resc_name, options)
    284         self._upload_collection(local_path, irods_path, overwrite, resc_name, options)
    285     else:
--> 286         self._obj_put(local_path, irods_path, overwrite, resc_name, options)
    287 except irods.exception.CUT_ACTION_PROCESSED_ERR as exc:
    288     raise irods.exception.CUT_ACTION_PROCESSED_ERR(
    289         f"During upload operation to '{irods_path}': iRODS server forbids action.") from exc

File ~/git-repos/iBridges/ibridges/irodsconnector/data_operations.py:135, in DataOperations._obj_put(self, local_path, irods_path, overwrite, resc_name, options)
    132     raise ValueError("local_path must be a file.")
    134 # Check if irods object already exists
--> 135 obj_exists = IrodsPath(self.session,
    136                        irods_path / local_path.name).dataobject_exists() \
    137              or irods_path.dataobject_exists()
    139 options = {
    140     kw.ALL_KW: '',
    141     kw.NUM_THREADS_KW: kw.NUM_THREADS,
    142     kw.REG_CHKSUM_KW: '',
    143     kw.VERIFY_CHKSUM_KW: ''
    144 }
    145 if resc_name not in ['', None]:

File ~/git-repos/iBridges/ibridges/utils/path.py:18, in IrodsPath.__init__(self, session, *args)
     16 def __init__(self, session, *args):
     17     self.session = session
---> 18     self._path = PurePosixPath(*args)
     19     super().__init__()

File /opt/homebrew/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pathlib.py:562, in PurePath.__new__(cls, *args)
    560 if cls is PurePath:
    561     cls = PureWindowsPath if os.name == 'nt' else PurePosixPath
--> 562 return cls._from_parts(args)

File /opt/homebrew/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pathlib.py:594, in PurePath._from_parts(cls, args)
    589 @classmethod
    590 def _from_parts(cls, args):
    591     # We need to call _parse_args on the instance, so as to get the
    592     # right flavour.
    593     self = object.__new__(cls)
--> 594     drv, root, parts = self._parse_args(args)
    595     self._drv = drv
    596     self._root = root

File /opt/homebrew/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pathlib.py:578, in PurePath._parse_args(cls, args)
    576     parts += a._parts
    577 else:
--> 578     a = os.fspath(a)
    579     if isinstance(a, str):
    580         # Force-cast str subclasses to str (issue #21127)
    581         parts.append(str(a))

TypeError: expected str, bytes or os.PathLike object, not IrodsPath
```

File: ibridges/utils/path.py (hunk 4 above)
Comment (qubixes) [reply ibridges/utils/path.py:125]: I'm going to throw an error instead if you don't mind!

File: ibridges/utils/path.py (hunk 3 above)
Comment (qubixes) [reply ibridges/utils/path.py:101]: Same as with the walk method.

File: ibridges/irodsconnector/session.py (hunk 1 above)
Comment (qubixes) [reply ibridges/irodsconnector/session.py:56]: Done!

File: tests/test_irodspath.py (hunk 5 above)
Comment (qubixes) [reply tests/test_irodspath.py]: No, I removed it.

**Comment by qubixes** [comment]
@chStaiger I have tried to fix that particular error. Can you see if it works again for you? From my part this PR is ready for merging.

**Pull request merged**